File: Trees/LC_1466.py
class Solution:
    def minReorder(self, n: int, connections: List[List[int]]) -> int:
        graph = collections.defaultdict(list)
        seen = set()
        for u, v in connections:
            seen.add((u,v))
            graph[u].append(v)
            graph[v].append(u)
        
        q = collections.deque()
        q.append(0)
        visited = set()
        visited.add(0)

        ans = 0

        while q:
            size = len(q)
            for _ in range(size):
                cur = q.popleft()
                for nxt in graph[cur]:
                    if nxt in visited:
                        continue
                    if (cur, nxt) in seen: # reverse the edge
                        ans += 1
                    q.append(nxt)
                    visited.add(nxt)
        return ans        

File: Trees/LC_1466.cpp
class Solution {
private:
    int ans;
public:
    void dfs(int cur, unordered_map<int, vector<int>>& graph, unordered_map<int, unordered_map<int, bool>>& edge, unordered_set<int>& visited){
        visited.insert(cur);
        for (int nxt: graph[cur]){
            if (visited.count(nxt)){
                continue;
            }
            if (edge.count(cur) && edge[cur].count(nxt)){
                ans++;
            }
            dfs(nxt, graph, edge, visited);
        }
        return;
    }
    int minReorder(int n, vector<vector<int>>& connections) {
        unordered_map<int, vector<int>> graph;
        unordered_map<int, unordered_map<int, bool>> edge;
        ans = 0;
        for (vector<int> c: connections){
            edge[c[0]][c[1]] = true;
            graph[c[0]].push_back(c[1]);
            graph[c[1]].push_back(c[0]);
        }
        unordered_set<int> visited;
        dfs(0, graph, edge, visited);
        return ans;
    }
};

// BFS version
class Solution {
public:
    int minReorder(int n, vector<vector<int>>& connections) {
        unordered_map<int, vector<int>> graph;
        unordered_map<int, unordered_map<int, bool>> seen;
        for (vector<int> c: connections){
            seen[c[0]][c[1]] = true;
            graph[c[0]].push_back(c[1]);
            graph[c[1]].push_back(c[0]);
        }
        queue<int> q;
        q.push(0);
        unordered_set<int> visited;
        visited.insert(0);
        int ans = 0;
        while(!q.empty()){
            int size = q.size();
            for (int iter = 0; iter < size; iter++){
                int cur = q.front();
                q.pop();
                for (int nxt : graph[cur]){
                    if (visited.count(nxt)){
                        continue;
                    }
                    if (seen.count(cur) && seen[cur].count(nxt)){
                        ans += 1;
                    }
                    visited.insert(nxt);
                    q.push(nxt);
                }
            }
        }
        return ans;
    }
};